iguana.corecontroller/src/main/java/org/aksw/iguana/cc/utils/IndexedQueryReader.java

```diff
@@ -0,0 +1,165 @@
+
+    /** The file whose content should be indexed. */
+    private final File file;
+
+    /**
+     * Indexes each content in between two of the given separators (including the beginning and end of the file). The
+     * given separator isn't allowed to be blank.
+     * @param filepath path to the file
+     * @param separator the separator line that is used in the file (isn't allowed to be blank)
+     * @return reader to access the indexed content
+     * @throws IllegalArgumentException the given separator was blank
+     * @throws IOException
+     */
+    public static IndexedQueryReader makeWithStringSeparator(String filepath, String separator) throws IOException {
+        if(separator.isBlank())
+            throw new IllegalArgumentException("Separator for makeWithStringSeparator can not be blank.");
+        return new IndexedQueryReader(filepath, separator);
+    }
+
+    /**
+     * Indexes every bundle of lines inside the file, that are in between two empty lines (including the beginning and
+     * end of the file). <br/>
+     * It uses the doubled line ending of the file as a separator, for example "\n\n".
+     * @param filepath path to the file
+     * @return reader to access the indexed content
+     * @throws IOException
+     */
+    public static IndexedQueryReader makeWithEmptyLines(String filepath) throws IOException {
+        String lineEnding = FileUtils.getLineEnding(filepath);
+        return new IndexedQueryReader(filepath, lineEnding + lineEnding);
+    }
+
+    /**
+     * Indexes every non-blank line inside the given file. It uses the line ending of the file as a separator.
+     * @param filepath path to the file
+     * @return reader to access the indexed lines
+     * @throws IOException
+     */
+    public static IndexedQueryReader make(String filepath) throws IOException {
+        return new IndexedQueryReader(filepath, FileUtils.getLineEnding(filepath));
+    }
+
+    /**
+     * Creates an object that indexes each content in between two of the given separators (including the beginning and
+     * end of the given file). <br/>
+     * @param filepath path to the file
+     * @param separator the separator for each bundle
+     * @throws IOException
+     */
+    private IndexedQueryReader(String filepath, String separator) throws IOException {
+        this.file = new File(filepath);
+        this.indexFile(separator);
+    }
+
+    /**
+     * Returns the indexed content with the given index.
+     * @param index the index of the searched content
+     * @return the searched content
+     * @throws IOException
+     */
+    public String readQuery(int index) throws IOException {
+        // split long value into two integers
+        long length = this.indices.get(index)[1];
+        int upper = (int)(length >> 32);
+        int lower = (int)length;
+
+        // store content into two byte arrays (for the case that the content is >~2GB)
+        byte[] data1 = new byte[lower];
```

Comment:
In Java a string and a byte array share both the same size limit wound up by the int size type. So this approach won't work. The only option (viable) option I see to retrieve larger queries is to stream them directly from the file. For this PR I suggest to just assume for the moment that queries are smaller than 2GB :)